File: terminal-capture-shell-golang-pty-standalone/cmd/check.go
package cmd

import (
	"flag"
	"fmt"
	"os"
	"os/exec"

	"flashback-shell-pty/pkg/config"
	"flashback-shell-pty/pkg/log"
	"flashback-shell-pty/pkg/ptywrap"
)

// CheckCmd handles the "check" subcommand.
func CheckCmd(cfg *config.Config, src *config.ConfigWithSource, logger *log.Logger, args []string) int {
	fs := flag.NewFlagSet("check", flag.ExitOnError)
	_, err := config.ParseSubcommandFlags(fs, args)
	if err != nil {
		logger.Errorf("failed to parse flags: %v", err)
		return 1
	}

	fmt.Println("flashback-shell-pty configuration:")
	fmt.Printf("  server_url:          %s (%s)\n", src.ServerURL.Value, src.ServerURL.Origin)
	fmt.Printf("  shell:               %s (%s)\n", src.Shell.Value, src.Shell.Origin)
	fmt.Printf("  buffer_size:         %s (%s)\n", src.BufferSize.Value, src.BufferSize.Origin)
	fmt.Printf("  buffer_mode:         %s (%s)\n", src.BufferMode.Value, src.BufferMode.Origin)
	fmt.Printf("  buffer_dir:          %s (%s)\n", src.BufferDir.Value, src.BufferDir.Origin)
	fmt.Printf("  device_id:           %s (%s)\n", src.DeviceID.Value, src.DeviceID.Origin)
	fmt.Printf("  capture_interval:    %s (%s)\n", src.CaptureInterval.Value, src.CaptureInterval.Origin)
	fmt.Printf("  first_capture_delay: %s (%s)\n", src.FirstCaptureDelay.Value, src.FirstCaptureDelay.Origin)
	fmt.Printf("  disable_capture:     %s (%s)\n", src.DisableCapture.Value, src.DisableCapture.Origin)
	fmt.Printf("  diff_only:           %s (%s)\n", src.DiffOnly.Value, src.DiffOnly.Origin)
	fmt.Printf("  diff_mode:           %s (%s)\n", src.DiffMode.Value, src.DiffMode.Origin)
	fmt.Printf("  text_only:           %s (%s)\n", src.TextOnly.Value, src.TextOnly.Origin)
	fmt.Printf("  scrollback_lines:    %s (%s)\n", src.ScrollbackLines.Value, src.ScrollbackLines.Origin)
	fmt.Printf("  allow_nested:        %s (%s)\n", src.AllowNested.Value, src.AllowNested.Origin)

	fmt.Println()
	fmt.Println("Validation:")

	shellBin := cfg.Shell
	if shellBin == "" {
		shellBin = os.Getenv("SHELL")
		if shellBin == "" {
			shellBin = "/bin/bash"
		}
	}
	if _, err := os.Stat(shellBin); err != nil {
		fmt.Printf("  shell:             FAIL (%s not found: %v)\n", shellBin, err)
		return 1
	}
	if _, err := exec.LookPath(shellBin); err != nil {
		fmt.Printf("  shell:             FAIL (%s not executable: %v)\n", shellBin, err)
		return 1
	}
	fmt.Printf("  shell:             OK (%s)\n", shellBin)

	// Try to open a PTY with a trivial command.
	testCmd := exec.Command(shellBin, "-c", "exit 0")
	ptmx, err := ptywrap.Start(testCmd)
	if err != nil {
		fmt.Printf("  pty:               FAIL (%v)\n", err)
		return 1
	}
	_ = ptmx.Close()
	_ = testCmd.Wait()
	fmt.Println("  pty:               OK")

	return 0
}

File: terminal-capture-shell-rust-standalone/src/cmd/check.rs
use std::process::ExitCode;

use crate::config::{Config, ConfigSources};
use crate::log::Logger;
use crate::pty;

pub async fn run(
    config: Config,
    sources: ConfigSources,
    logger: Logger,
) -> ExitCode {
    println!("terminal-capture-shell-rust-standalone configuration:");
    print_source("server_url", &sources.server_url);
    print_source("shell", &sources.shell);
    print_source("buffer_size", &sources.buffer_size);
    print_source("buffer_mode", &sources.buffer_mode);
    print_source("buffer_dir", &sources.buffer_dir);
    print_source("device_id", &sources.device_id);
    print_source("capture_interval", &sources.capture_interval);
    print_source("first_capture_delay", &sources.first_capture_delay);
    print_source("disable_capture", &sources.disable_capture);
    print_source("diff_only", &sources.diff_only);
    print_source("diff_mode", &sources.diff_mode);
    print_source("text_only", &sources.text_only);
    print_source("scrollback_lines", &sources.scrollback_lines);
    print_source("allow_nested", &sources.allow_nested);

    println!();
    println!("Validation:");

    let shell_bin = config.shell_binary();
    if !std::path::Path::new(&shell_bin).exists() {
        println!("  shell:             FAIL ({shell_bin} not found)");
        return ExitCode::from(1);
    }
    if std::env::split_paths(&std::env::var("PATH").unwrap_or_default())
        .map(|p| p.join(&shell_bin))
        .any(|p| p.exists())
        || std::path::Path::new(&shell_bin).is_absolute()
    {
        // ok
    } else {
        println!("  shell:             FAIL ({shell_bin} not executable from PATH)");
        return ExitCode::from(1);
    }
    println!("  shell:             OK ({shell_bin})");

    // Try to open a PTY with a trivial command.
    let command = vec![shell_bin, "-c".to_owned(), "exit 0".to_owned()];
    let winsize = nix::pty::Winsize {
        ws_row: 24,
        ws_col: 80,
        ws_xpixel: 0,
        ws_ypixel: 0,
    };
    let extra_env = std::collections::HashMap::new();
    match pty::spawn(&command, winsize, &extra_env) {
        Ok(pty) => {
            let _ = pty.wait(None).await;
            println!("  pty:               OK");
        }
        Err(e) => {
            println!("  pty:               FAIL ({e})");
            return ExitCode::from(1);
        }
    }

    logger.info("check completed successfully");
    ExitCode::SUCCESS
}

fn print_source(name: &str, source: &crate::config::Source) {
    println!("  {name:19} {} ({})", source.value, source.origin);
}
